runtime-sdk/src/modules/core/mod.rs

```diff
@@ -225,6 +239,35 @@ impl module::AuthHandler for Module {
         }
         Ok(())
     }
+
+    fn before_handle_call<C: TxContext>(ctx: &mut C, _call: &Call) -> Result<(), Error> {
+        let mut num_signature: u64 = 0;
+        let mut num_multisig_signer: u64 = 0;
+        for si in &ctx.tx_auth_info().signer_info {
+            match &si.address_spec {
+                AddressSpec::Signature(_) => {
+                    num_signature = num_signature.checked_add(1).ok_or(Error::GasOverflow)?;
+                }
+                AddressSpec::Multisig(config) => {
+                    num_multisig_signer = num_multisig_signer
+                        .checked_add(config.signers.len() as u64)
+                        .ok_or(Error::GasOverflow)?;
+                }
+            }
+        }
+        let params = Self::params(ctx.runtime_state());
+        let total = num_signature
+            .checked_mul(params.gas_costs.auth_signature)
+            .ok_or(Error::GasOverflow)?
+            .checked_add(
+                num_multisig_signer
+                    .checked_mul(params.gas_costs.auth_multisig_signer)
+                    .ok_or(Error::GasOverflow)?,
+            )
+            .ok_or(Error::GasOverflow)?;
```

Comment:
```suggestion
        let total = num_signature
            .checked_mul(params.gas_costs.auth_signature)
            .and_then(|v| {
                v.checked_add(num_multisig_signer.checked_mul(params.gas_costs.auth_multisig_signer)?)
            })
            .ok_or(Error::GasOverflow)?;
```
and then you only have to do error handling once at the very end

Reply:
hm now I'm wondering if we can go further with `?`s if we put most of the logic into a lambda